File: samples/java/ZoneWorld/Client/src/main/java/systems/zlink/samples/zoneworld/client/ScenarioSupport.java
package systems.zlink.samples.zoneworld.client;

import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldSpec;
import systems.zlink.stream.connector.ZLinkStreamConnector;
import systems.zlink.stream.connector.ZLinkStreamConnectorFactory;
import systems.zlink.stream.connector.ZLinkStreamConnectorOptions;
import systems.zlink.stream.connector.ZLinkStreamDispatchMode;
import systems.zlink.stream.connector.ZLinkStreamMessage;

import java.net.URI;
import java.time.Duration;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Predicate;

final class ScenarioSupport {
    static final Duration REQUEST_TIMEOUT = Duration.ofSeconds(15);
    static final Duration CROSS_NODE_TIMEOUT = Duration.ofSeconds(30);

    private ScenarioSupport() {}

    static ZLinkStreamConnector connect(String endpoint) {
        ZLinkStreamConnector connector =
                ZLinkStreamConnectorFactory.create(
                        new ZLinkStreamConnectorOptions(
                                URI.create(endpoint),
                                ZLinkStreamDispatchMode.IMMEDIATE,
                                REQUEST_TIMEOUT,
                                REQUEST_TIMEOUT,
                                2,
                                Duration.ofSeconds(5),
                                64 * 1024,
                                64 * 1024,
                                true,
                                Duration.ofSeconds(1),
                                Duration.ofSeconds(5),
                                true,
                                Duration.ofMillis(250),
                                Duration.ofSeconds(5),
                                2.0,
                                false,
                                null,
                                null,
                                null,
                                null));
        connector.connect().submit().toCompletableFuture().join();
        return connector;
    }

    static <T> CompletionStage<ZLinkStreamMessage<T>> waitFor(
            ZLinkStreamConnector connector,
            Class<T> type,
            Predicate<T> predicate,
            Duration timeout) {
        return connector
                .waitFor(type)
                .where(type, message -> predicate.test(message.payload()))
                .timeout(timeout)
                .submit(type);
    }

    static <T> T request(ZLinkStreamConnector connector, Object request, Class<T> reply) {
        return connector
                .request(request)
                .timeout(REQUEST_TIMEOUT)
                .submit(reply)
                .toCompletableFuture()
                .join();
    }

    static void delay(Duration duration) {
        CompletableFuture.runAsync(
                        () -> {},
                        CompletableFuture.delayedExecutor(
                                duration.toMillis(), TimeUnit.MILLISECONDS))
                .join();
    }

    static void close(ZLinkStreamConnector connector) {
        if (connector != null) connector.close().submit().toCompletableFuture().join();
    }

    static void ensure(boolean condition, String message) {
        if (!condition) throw new IllegalStateException(message);
    }

    static String unique(String prefix) {
        return prefix
                + "-"
                + java.util.UUID.randomUUID().toString().replace("-", "").substring(0, 6);
    }

    static Point center(String zone) {
        return switch (zone) {
            case "zone-nw" -> new Point(25, 25);
            case "zone-ne" -> new Point(75, 25);
            case "zone-sw" -> new Point(25, 75);
            case "zone-se" -> new Point(75, 75);
            default -> throw new IllegalArgumentException("unknown zone " + zone);
        };
    }

    static Edge edge(String source, String target) {
        return switch (source + ":" + target) {
            case "zone-nw:zone-ne" -> new Edge(new Point(48, 25), new Point(52, 25));
            case "zone-nw:zone-sw" -> new Edge(new Point(25, 48), new Point(25, 52));
            case "zone-ne:zone-se" -> new Edge(new Point(75, 48), new Point(75, 52));
            case "zone-sw:zone-se" -> new Edge(new Point(48, 75), new Point(52, 75));
            case "zone-ne:zone-nw" -> new Edge(new Point(52, 25), new Point(48, 25));
            case "zone-sw:zone-nw" -> new Edge(new Point(25, 52), new Point(25, 48));
            case "zone-se:zone-ne" -> new Edge(new Point(75, 52), new Point(75, 48));
            case "zone-se:zone-sw" -> new Edge(new Point(52, 75), new Point(48, 75));
            default -> throw new IllegalArgumentException("zones are not adjacent");
        };
    }

    record Point(int x, int y) {}

    record Edge(Point source, Point target) {}

    static final class Game implements AutoCloseable {
        final ZLinkStreamConnector connector;
        final String playerId;
        int x;
        int y;

        Game(ClientOptions options, String playerId) {
            this.connector = connect(options.gatewayEndpoint());
            this.playerId = playerId;
        }

        Messages.JoinWorldNotify join() {
            CompletionStage<ZLinkStreamMessage<Messages.JoinWorldNotify>> response =
                    waitFor(
                            connector,
                            Messages.JoinWorldNotify.class,
                            value -> playerId.equals(value.playerId()),
                            Duration.ofSeconds(20));
            connector
                    .send(new Messages.JoinWorldMsg(playerId))
                    .submit()
                    .toCompletableFuture()
                    .join();
            Messages.JoinWorldNotify value = response.toCompletableFuture().join().payload();
            x = value.x();
            y = value.y();
            return value;
        }

        void move(int targetX, int targetY) {
            connector
                    .send(new Messages.MoveMsg(targetX, targetY))
                    .submit()
                    .toCompletableFuture()
                    .join();
        }

        Messages.ZoneChangedNotify moveTo(int targetX, int targetY) {
            Messages.ZoneChangedNotify last = null;
            while (x != targetX || y != targetY) {
                int nextX =
                        x != targetX
                                ? x
                                        + Math.clamp(
                                                targetX - x,
                                                -ZoneWorldSpec.MAX_STEP_PER_AXIS,
                                                ZoneWorldSpec.MAX_STEP_PER_AXIS)
                                : x;
                int nextY =
                        x == targetX
                                ? y
                                        + Math.clamp(
                                                targetY - y,
                                                -ZoneWorldSpec.MAX_STEP_PER_AXIS,
                                                ZoneWorldSpec.MAX_STEP_PER_AXIS)
                                : y;
                String oldZone = ZoneWorldSpec.zoneOf(x, y);
                String newZone = ZoneWorldSpec.zoneOf(nextX, nextY);
                if (!oldZone.equals(newZone)) {
                    CompletionStage<ZLinkStreamMessage<Messages.ZoneChangedNotify>> changed =
                            waitFor(
                                    connector,
                                    Messages.ZoneChangedNotify.class,
                                    value ->
                                            playerId.equals(value.playerId())
                                                    && newZone.equals(value.zoneId()),
                                    CROSS_NODE_TIMEOUT);
                    move(nextX, nextY);
                    last = changed.toCompletableFuture().join().payload();
                } else {
                    CompletionStage<ZLinkStreamMessage<Messages.ZoneStateNotify>> arrived =
                            waitFor(
                                    connector,
                                    Messages.ZoneStateNotify.class,
                                    state ->
                                            state.players().stream()
                                                    .anyMatch(
                                                            player ->
                                                                    playerId.equals(
                                                                                    player
                                                                                            .playerId())
                                                                            && player.x() == nextX
                                                                            && player.y() == nextY),
                                    CROSS_NODE_TIMEOUT);
                    move(nextX, nextY);
                    arrived.toCompletableFuture().join();
                }
                x = nextX;
                y = nextY;
            }
            return last;
        }

        @Override
        public void close() {
            ScenarioSupport.close(connector);
        }
    }

    static final class Ops implements AutoCloseable {
        final ZLinkStreamConnector connector;

        Ops(ClientOptions options) {
            connector = connect(options.opsEndpoint());
        }

        Messages.WatchNodesRes watch() {
            return request(connector, new Messages.WatchNodesReq(), Messages.WatchNodesRes.class);
        }

        Messages.SetMaintenanceRes maintenance(String nodeId, boolean enabled) {
            CompletionStage<ZLinkStreamMessage<Messages.NodeStatusNotify>> observed =
                    waitFor(
                            connector,
                            Messages.NodeStatusNotify.class,
                            value ->
                                    nodeId.equals(value.nodeId()) && value.maintenance() == enabled,
                            Duration.ofSeconds(20));
            Messages.SetMaintenanceRes result =
                    request(
                            connector,
                            new Messages.SetMaintenanceReq(nodeId, enabled),
                            Messages.SetMaintenanceRes.class);
            if (result.error() == null) observed.toCompletableFuture().join();
            return result;
        }

        @Override
        public void close() {
            ScenarioSupport.close(connector);
        }
    }

    static final class Probes implements AutoCloseable {
        final ZLinkStreamConnector connector;

        Probes(ClientOptions options) {
            connector = connect(options.gatewayEndpoint());
        }

        Messages.RelocationPairRes pair() {
            return request(
                    connector, new Messages.RelocationPairReq(), Messages.RelocationPairRes.class);
        }

        Messages.ActorLocationProbeRes actor(String id) {
            return request(
                    connector,
                    new Messages.ActorLocationProbeReq(id),
                    Messages.ActorLocationProbeRes.class);
        }

        Messages.FreshActorProbeRes fresh(String id) {
            return request(
                    connector,
                    new Messages.FreshActorProbeReq(id),
                    Messages.FreshActorProbeRes.class);
        }

        Messages.MessageFollowProbeRes probe(String actor, String id, byte[] payload) {
            return request(
                    connector,
                    new Messages.MessageFollowProbeReq(actor, id, payload),
                    Messages.MessageFollowProbeRes.class);
        }

        void sendProbe(String actor, String id, byte[] payload) {
            connector
                    .send(new Messages.MessageFollowProbeMsg(actor, id, payload))
                    .submit()
                    .toCompletableFuture()
                    .join();
        }

        @Override
        public void close() {
            ScenarioSupport.close(connector);
        }
    }
}

File: samples/kotlin/ZoneWorld/Client/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/client/ScenarioSupport.kt
package systems.zlink.samples.kotlin.zoneworld.client

import java.net.URI
import java.time.Duration
import java.util.UUID
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import systems.zlink.framework.kotlin.ZLinkKotlinStreamConnector
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.awaitReply
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldSpec
import systems.zlink.stream.connector.ZLinkStreamConnectorFactory
import systems.zlink.stream.connector.ZLinkStreamConnectorOptions
import systems.zlink.stream.connector.ZLinkStreamDispatchMode

internal val REQUEST_TIMEOUT: Duration = Duration.ofSeconds(15)
internal val CROSS_NODE_TIMEOUT: Duration = Duration.ofSeconds(30)

internal suspend fun connect(endpoint: String): ZLinkKotlinStreamConnector {
    val connector =
        ZLinkStreamConnectorFactory.create(
                ZLinkStreamConnectorOptions(
                    URI.create(endpoint),
                    ZLinkStreamDispatchMode.IMMEDIATE,
                    REQUEST_TIMEOUT,
                    REQUEST_TIMEOUT,
                    2,
                    Duration.ofSeconds(5),
                    64 * 1024,
                    64 * 1024,
                    true,
                    Duration.ofSeconds(1),
                    Duration.ofSeconds(5),
                    true,
                    Duration.ofMillis(250),
                    Duration.ofSeconds(5),
                    2.0,
                    false,
                    null,
                    null,
                    null,
                    null,
                )
            )
            .kotlin()
    connector.connect().await()
    return connector
}

internal fun ensure(condition: Boolean, message: String) {
    check(condition) { message }
}

internal fun unique(prefix: String) =
    "$prefix-${UUID.randomUUID().toString().replace("-", "").take(6)}"

internal data class Point(val x: Int, val y: Int)

internal data class Edge(val source: Point, val target: Point)

internal fun center(zone: String) =
    when (zone) {
        "zone-nw" -> Point(25, 25)
        "zone-ne" -> Point(75, 25)
        "zone-sw" -> Point(25, 75)
        "zone-se" -> Point(75, 75)
        else -> error("unknown zone $zone")
    }

internal fun edge(source: String, target: String) =
    when ("$source:$target") {
        "zone-nw:zone-ne" -> Edge(Point(48, 25), Point(52, 25))
        "zone-nw:zone-sw" -> Edge(Point(25, 48), Point(25, 52))
        "zone-ne:zone-se" -> Edge(Point(75, 48), Point(75, 52))
        "zone-sw:zone-se" -> Edge(Point(48, 75), Point(52, 75))
        "zone-ne:zone-nw" -> Edge(Point(52, 25), Point(48, 25))
        "zone-sw:zone-nw" -> Edge(Point(25, 52), Point(25, 48))
        "zone-se:zone-ne" -> Edge(Point(75, 52), Point(75, 48))
        "zone-se:zone-sw" -> Edge(Point(52, 75), Point(48, 75))
        else -> error("zones are not adjacent")
    }

internal interface AsyncCloseable {
    suspend fun close()
}

internal suspend fun <T> withResources(
    vararg resources: AsyncCloseable,
    block: suspend () -> T,
): T =
    try {
        block()
    } finally {
        resources.reversed().forEach { runCatching { it.close() } }
    }

internal class Game
private constructor(val connector: ZLinkKotlinStreamConnector, val playerId: String) :
    AsyncCloseable {
    var x = 0
    var y = 0

    suspend fun join(): Messages.JoinWorldNotify = coroutineScope {
        val response =
            async(start = CoroutineStart.UNDISPATCHED) {
                connector
                    .waitFor<Messages.JoinWorldNotify>()
                    .where { it.payload().playerId == playerId }
                    .timeout(Duration.ofSeconds(20))
                    .await()
                    .payload()
            }
        connector.send(Messages.JoinWorldMsg(playerId)).await()
        response.await().also {
            x = it.x
            y = it.y
        }
    }

    suspend fun move(targetX: Int, targetY: Int) {
        connector.send(Messages.MoveMsg(targetX, targetY)).await()
    }

    suspend fun moveTo(targetX: Int, targetY: Int): Messages.ZoneChangedNotify? = coroutineScope {
        var last: Messages.ZoneChangedNotify? = null
        while (x != targetX || y != targetY) {
            val nextX =
                if (x != targetX)
                    x +
                        (targetX - x).coerceIn(
                            -ZoneWorldSpec.MAX_STEP_PER_AXIS,
                            ZoneWorldSpec.MAX_STEP_PER_AXIS,
                        )
                else x
            val nextY =
                if (x == targetX)
                    y +
                        (targetY - y).coerceIn(
                            -ZoneWorldSpec.MAX_STEP_PER_AXIS,
                            ZoneWorldSpec.MAX_STEP_PER_AXIS,
                        )
                else y
            val oldZone = ZoneWorldSpec.zoneOf(x, y)
            val newZone = ZoneWorldSpec.zoneOf(nextX, nextY)
            if (oldZone != newZone) {
                val changed =
                    async(start = CoroutineStart.UNDISPATCHED) {
                        connector
                            .waitFor<Messages.ZoneChangedNotify>()
                            .where {
                                it.payload().playerId == playerId && it.payload().zoneId == newZone
                            }
                            .timeout(CROSS_NODE_TIMEOUT)
                            .await()
                            .payload()
                    }
                move(nextX, nextY)
                last = changed.await()
            } else {
                val arrived =
                    async(start = CoroutineStart.UNDISPATCHED) {
                        connector
                            .waitFor<Messages.ZoneStateNotify>()
                            .where { state ->
                                state.payload().players.any {
                                    it.playerId == playerId && it.x == nextX && it.y == nextY
                                }
                            }
                            .timeout(CROSS_NODE_TIMEOUT)
                            .await()
                    }
                move(nextX, nextY)
                arrived.await()
            }
            x = nextX
            y = nextY
        }
        last
    }

    override suspend fun close() = connector.close().await()

    companion object {
        suspend fun create(options: ClientOptions, playerId: String) =
            Game(connect(options.gatewayEndpoint), playerId)
    }
}

internal class Ops private constructor(val connector: ZLinkKotlinStreamConnector) : AsyncCloseable {
    suspend fun watch(): Messages.WatchNodesRes =
        connector.request(Messages.WatchNodesReq()).timeout(REQUEST_TIMEOUT).awaitReply()

    suspend fun maintenance(nodeId: String, enabled: Boolean): Messages.SetMaintenanceRes =
        coroutineScope {
            val observed =
                async(start = CoroutineStart.UNDISPATCHED) {
                    connector
                        .waitFor<Messages.NodeStatusNotify>()
                        .where {
                            it.payload().nodeId == nodeId && it.payload().maintenance == enabled
                        }
                        .timeout(Duration.ofSeconds(20))
                        .await()
                }
            val result =
                connector
                    .request(Messages.SetMaintenanceReq(nodeId, enabled))
                    .timeout(REQUEST_TIMEOUT)
                    .awaitReply<Messages.SetMaintenanceRes>()
            if (result.error == null) observed.await() else observed.cancel()
            result
        }

    override suspend fun close() = connector.close().await()

    companion object {
        suspend fun create(options: ClientOptions) = Ops(connect(options.opsEndpoint))
    }
}

internal class Probes private constructor(val connector: ZLinkKotlinStreamConnector) :
    AsyncCloseable {
    suspend fun pair(): Messages.RelocationPairRes =
        connector.request(Messages.RelocationPairReq()).timeout(REQUEST_TIMEOUT).awaitReply()

    suspend fun actor(id: String): Messages.ActorLocationProbeRes =
        connector.request(Messages.ActorLocationProbeReq(id)).timeout(REQUEST_TIMEOUT).awaitReply()

    suspend fun fresh(id: String): Messages.FreshActorProbeRes =
        connector.request(Messages.FreshActorProbeReq(id)).timeout(REQUEST_TIMEOUT).awaitReply()

    suspend fun probe(
        actor: String,
        id: String,
        payload: ByteArray,
    ): Messages.MessageFollowProbeRes =
        connector
            .request(Messages.MessageFollowProbeReq(actor, id, payload))
            .timeout(REQUEST_TIMEOUT)
            .awaitReply()

    suspend fun sendProbe(actor: String, id: String, payload: ByteArray) =
        connector.send(Messages.MessageFollowProbeMsg(actor, id, payload)).await()

    override suspend fun close() = connector.close().await()

    companion object {
        suspend fun create(options: ClientOptions) = Probes(connect(options.gatewayEndpoint))
    }
}
